python/ray/util/sgd/torch/worker_group.py

```diff
@@ -0,0 +1,480 @@
+            [
+                worker.__ray_terminate__.remote()
+                for worker in self.remote_workers
+            ]
+        except RayActorError:
+            logger.warning("Failed to shutdown gracefully, forcing a "
+                           "shutdown.")
+            self._reset()
+
+    def shutdown(self, force=False):
+        if not force:
+            cleanup = [
+                worker.shutdown.remote() for worker in self.remote_workers
+            ]
+            self._terminate_remote_workers(cleanup)
+        else:
+            self._reset()
+        self.remote_workers = []
+
+    def _reset(self):
+        for worker in self.remote_workers:
+            logger.debug(f"Killing worker {worker}.")
+            ray.kill(worker)
+        self.remote_workers = []
+
+    def should_resize(self):
+        """Returns True if past cooldown and exists resources to scale up."""
+        worker_gap = self.num_workers - len(self.remote_workers)
+        past_cooldown = (time.time() - self._last_resize) > RESIZE_COOLDOWN_S
+        if past_cooldown and worker_gap:
+            # Assume 1 resource is already reserved for local worker.
+            potential_remote_size = self._check_potential_remote_workers_size()
+            return potential_remote_size > 0
+        return False
+
+    def _check_potential_remote_workers_size(self):
+        # ASSUME 1 GPU + 1 CPU is already reserved for the local worker
+        remote_resources = ray.available_resources()
+        max_remote_workers = self.num_workers
+        new_remote_workers = min(
+            remote_resources.get("CPU", 0), max_remote_workers)
+        if self.use_gpu:
+            new_remote_workers = min(
+                remote_resources.get("GPU", 0), new_remote_workers)
+        return new_remote_workers
+
+    def resize_workers(self, max_retries=10):
+        self._reset()
+
+        time.sleep(1)
+        for i in range(max_retries):
+            new_remote_workers = self._check_potential_remote_workers_size()
+            if new_remote_workers:
+                self._last_resize = time.time()
+                self.start_workers(
+                    int(new_remote_workers),
+                    params=self.params,
+                    dist_params=self.dist_params,
+                    initialization_hook=self.initialization_hook,
+                    num_cpus_per_worker=self.num_cpus_per_worker,
+                    use_gpu=self.use_gpu,
+                    timeout_s=self.timeout_s)
+                self.load_state_dict(self.state_dict())
+                return
+            else:
+                delay = 2**i
+                logger.warning(
+                    "No new workers found. Retrying in %d sec." % delay)
+                time.sleep(delay)
+        raise RuntimeError("Exceeded max_retries for relaunching workers.")
+
+    def get_num_workers(self):
+        return len(self.remote_workers)
+
+
+class LocalWorkerGroup(BaseWorkerGroup):
```

Comment:
Can you leave a note about why this is different from the RemoteWorkerGroup (as a docstring)?

Reply:
done